### ISSUE TITLE
IPv8/Dispersy dual stack peer explosion

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#4171]
Tribler takes a whole core for nearly 100% after running for days (167h of CPU of htop equal 75% avg. load):
![screenshot from 2019-01-26 17-41-05](https://user-images.githubusercontent.com/325224/51790171-f1aafc00-2191-11e9-8aec-3f5c2baa1680.png)
![screenshot from 2019-01-26 17-46-56](https://user-images.githubusercontent.com/325224/51790219-6d0cad80-2192-11e9-92b8-ff0b8cc24ee8.png)
The swp is getting full, maxed out box, but that does not boost user-space CPU usage. Running v7.1.5 mostly idle for 9 days continuously on an Ubuntu 18.04.1 LTS box. Never done a single keyword search, two Fedora seeding (0MB uploaded), but donating encrypted relay bandwidth:
![screenshot from 2019-01-26 17-33-50](https://user-images.githubusercontent.com/325224/51790054-a93f0e80-2190-11e9-9c9b-1257719be706.png)
Gathering solid credits:
![screenshot from 2019-01-26 17-37-25](https://user-images.githubusercontent.com/325224/51790086-210d3900-2191-11e9-9dd9-37526e79810e.png)
![screenshot from 2019-01-26 17-39-23](https://user-images.githubusercontent.com/325224/51790138-ad1f6080-2191-11e9-9488-64556604a360.png)

Reasonable amount of memory, but not growing out of control:
![screenshot from 2019-01-26 17-35-55](https://user-images.githubusercontent.com/325224/51790068-ee634080-2190-11e9-9109-177fd07977ad.png)

Some amount of errors in the logs:

```
INFO    1548520097.02                 community:83   (Request)  Timeout for ping to Peer<94.174.45.58:7759, pnK7Ahoxt5xGR/w29hfI5EATiZI=>
INFO    1548520097.70                   handler:152  (TftpHandler)  TFTP[23593 C 12.124.121.166:7759][8b6cd3e1e5dc97efc3cb469631d4980f6dfe13d3.torrent] timed out
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
WARNING 1548520098.09                  endpoint:56   (UDPEndpoint)  Dropping packet due to socket error: [Errno 11] Resource temporarily unavailable
INFO    1548520098.94                   handler:137  (TftpHandler)  TFTP[34101 C 23.252.56.68:7759][8b6cd3e1e5dc97efc3cb469631d4980f6dfe13d3.torrent] started
INFO    1548520101.04           torrent_checker:151  (TorrentChecker)  Selected 2 new torrents to check on tracker: udp://tracker.xfsub.com:6868
INFO    1548520101.07                   handler:152  (TftpHandler)  TFTP[34101 C 23.252.56.68:7759][8b6cd3e1e5dc97efc3cb469631d4980f6dfe13d3.torrent] timed out
INFO    1548520101.17                   session:419  (UdpTrackerSession)  Error when querying UDP tracker: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.DNSLookupError'>: DNS lookup failed: tracker.xfsub.com.
] udp://tracker.xfsub.com:6868
WARNING 1548520101.17           torrent_checker:262  (TorrentChecker)  Got session error for URL udp://tracker.xfsub.com:6868: [Failure instance: Traceback (failure with no frames): <type 'exceptions.ValueError'>: UDP tracker failed for url udp://tracker.xfsub.com:6868 (error: DNS lookup failed: tracker.xfsub.com.)
]
INFO    1548520102.49                   handler:137  (TftpHandler)  TFTP[1609 C 74.66.207.23:9683][8b6cd3e1e5dc97efc3cb469631d4980f6dfe13d3.torrent] started
WARNING 1548520103.63                   handler:240  (TftpHandler)  got non-existing session from 2.24.234.96:7759, id = 28608
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<86.126.173.209:19811, F8TPp56do0IvYd8QJ1KyoHijOMg=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<193.165.189.6:26423, rfJDpWM47gT55U9Izwe9R8tGROA=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<5.79.71.187:35035, hiNKfy4uEPKMH9IJ177OKc3zh3c=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<88.21.114.167:7759, s8Ro19O1qVHHvAnAjvAv1IAvbGw=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<70.121.56.5:7759, sLxOpSmBdM0YfABKCIHyZVuAxhk=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<92.20.137.112:55502, r8nBAHqUi9ezNc9TZZHMvTRcrVM=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<81.171.8.17:35200, 7s7HWCaTpTTAi6LSG09/EfZVxRo=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<81.171.8.19:35200, qIqJyrGyUl6VPJIBPmx2FBIZ3Ao=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<81.171.8.18:35160, XO8ZrfspX3GBzL89CMTRQ0YstB8=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<81.171.8.19:35075, PAOwFnIPqursajEC4nwLZgH2Tzw=>
INFO    1548520103.96                 community:83   (Request)  Timeout for ping to Peer<88.89.71.78:7758, o4/QqaweAuzAW85zt/HUQBfQ024=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<185.5.165.113:47892, rWrfbl3Y2xZaRXwwuZk9ZFYjd0I=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<81.171.8.17:35060, vAvMRdf0hPb5xeJ/PTviWHiQHKw=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<81.171.8.17:35010, KLP/0BIzlDf7HiYL2QpV+7oCWHk=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<81.171.27.193:35040, RAVrUhLCsiUUKx+ipKxmX2uV2lQ=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<81.171.27.193:35005, kcfBnK+3zeEGwy8J0LcWurCoZlg=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<37.48.77.235:35010, t07Ao4YBTvmvw6UqH9O7yLx5Mg0=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<37.48.77.237:35035, N+LrqteDPqm2fSi77BOudycsRwI=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<37.48.77.238:35020, E73U8Eh5g1vFIXbMmOq+NqplIRQ=>
INFO    1548520103.97                 community:83   (Request)  Timeout for ping to Peer<95.96.66.166:7759, uNOlxm1LhnBFxlRhwHLzrgG1zQM=>
INFO    1548520103.97                 community:83   (Request)  Timeout for find to Peer<86.126.173.209:19811, F8TPp56do0IvYd8QJ1KyoHijOMg=>
INFO    1548520104.89                   handler:152  (TftpHandler)  TFTP[1609 C 74.66.207.23:9683][8b6cd3e1e5dc97efc3cb469631d4980f6dfe13d3.torrent] timed out
INFO    1548520105.00                   handler:137  (TftpHandler)  TFTP[39475 C 75.169.51.124:10054][8b6cd3e1e5dc97efc3cb469631d4980f6dfe13d3.torrent] started
```

**Comment by synctext** [comment]
:fearful: :dizzy_face: :fearful: 2447 peers, what does that even mean?
![screenshot from 2019-01-26 19-05-59](https://user-images.githubusercontent.com/325224/51791056-aac30380-219d-11e9-9538-b683f3e7609e.png)

**Comment by devos50** [comment]
Wow, 2447 peers is clearly not ok. If I remember correctly, the number of peers in a community should be capped, especially for non-discovery communities. The `max_peers` parameters when initializing a community should indicate the maximum number of peers that are in a specific community.

Having so many open connections to others could lead to unexpected behavior. For example, having many peers in the DHT community might make you an attractive target for value lookups, essentially DDOS-ing you. It also seems that the system ran out of file descriptors: `[Errno 11] Resource temporarily unavailable`. @synctext are there any running downloads?

I think this is a high priority issue that should be fixed before releasing 7.1.6.

**Comment by qstokkink** [comment]
> 😨 😵 😨 2447 peers, what does that even mean?

You're keeping 2447 connections to others open.

All peers you gather over 20 are introductions by others, you no longer perform walks yourself at that point. Even by just passively being introduced by others, you have managed to gather quite the following.

To solve this we would need to start kicking out peers from communities. Which in turn may lead to problems with IPv8 deciding to kick someone out you were performing a transaction with (for example in the market).

**Comment by devos50** [comment]
I just created the 7.1.6 milestone and assigned some open PRs/issues to it which I think should be included in the release.

**Comment by devos50** [comment]
@synctext yes we should work towards a large-scale GigaChannel experiment, where we deploy 1000 Tribler instances on the DAS5 which communicate and synchronize channels/torrents with each other. This should be a nightly job. Getting this up and running might take some time, however.

**Comment by synctext** [comment]
30 sockets opened by core..
![screenshot from 2019-01-27 11-31-59](https://user-images.githubusercontent.com/325224/51799875-49e00d80-2227-11e9-8be6-3eab7db520d7.png)
![screenshot from 2019-01-27 11-31-17](https://user-images.githubusercontent.com/325224/51799876-49e00d80-2227-11e9-8f1b-86354462ee84.png)

**Comment by devos50** [comment]
@synctext in fact, this number shows the total number of open file descriptors, which also includes Python 2.7 library files and database files which are opened for reading. You should see them if you expand the view. Also, there are many processes in Tribler which try to open connections to remote peers (DHT lookups, libtorrent, TFTP, Dispersy, IPv8 etc).

**Comment by qstokkink** [comment]
Should be solved with the next IPv8 pointer update.

**Comment by synctext** [comment]
Sorry, overlay is more broken then we maybe realised. After 23hour of uptime on Mac:
![img_20190128_184048](https://user-images.githubusercontent.com/325224/51861554-d9c3ab80-233c-11e9-8270-09817006ba5d.jpg)

**Comment by devos50** [comment]
@synctext it seems that you are running an older version of Tribler (7.1.4), which does not contain the IPv8 fixes that (should) resolve this issue. You could try to install Tribler from this build (v7.2.0-rc1): https://jenkins-ci.tribler.org/job/Build-Tribler_release/job/Build-Custom/1/ and check whether it correctly resolves the issue?

**Comment by devos50** [comment]
Also, I will integrate an IPv8 peer count monitor in the application tester, so we can get overlay statistics over longer periods of time 👍

**Comment by synctext** [comment]
Running Ubuntu 7.1.5 with a clean megacache for 12hours reproduces this issue. We need a Gumby test for live-edge walking to design a load balancing algorithm:
![screenshot from 2019-01-29 09-52-20](https://user-images.githubusercontent.com/325224/51895917-981e1980-23ab-11e9-9a49-9ca7c88b3561.png)

**Comment by qstokkink** [comment]
This was only fixed in 7.1.6/7.2.0 [as @devos50 pointed out](https://github.com/Tribler/tribler/issues/4171#issuecomment-458297766), please reproduce this using the latest version. Also, I sincerely doubt this is due to the live edge walker.

**Comment by devos50** [comment]
The number of peers in each community is now monitored and plotted during the application tests. This should help us to debug this kind of overlay behavior.

![ipv8_overlays](https://user-images.githubusercontent.com/1707075/51904596-54cda600-23bf-11e9-9129-9315c83d6834.png)

**Comment by synctext** [comment]
This is now the last open issue before we can release V7.2. ToDo:
- [ ] find root-cause:
  - what is going on?
  - live edge walker
  - shared peer pool
- [ ] remove the temporary no-response fix (which impact DHT performance).
- [ ] repair with load balancing or something nice.

**Comment by qstokkink** [comment]
> - remove the temporary no-response fix (which impact DHT performance).
> - repair with load balancing or something nice.

I cannot disagree more. This is analogous to (1) unlocking the doors and (2) educating everyone on how bad crime is for society.

Load balancing is not a solution to counter attacks by malicious nodes. This "temporary fix" is our only line of defense against being completely DOS-attacked to death.

**Comment by synctext** [comment]
7.2.0RC1 has an overlay bug. Statistics are 20minutes uptime:
![screenshot from 2019-01-29 18-40-23](https://user-images.githubusercontent.com/325224/51928083-5c0fa680-23f5-11e9-9e50-6d339d808f6a.png)

**Comment by qstokkink** [comment]
@synctext that's the expected healthy peer count.

**Comment by devos50** [comment]
@qstokkink @synctext so are we going to postpone this issue to 7.3?

**Comment by qstokkink** [comment]
Well the original issue seems under control now. We can make a new issue to explore other means of walking and load balancing and close this one.

**Comment by synctext** [comment]
@qstokkink Do we still have a (performance) bug?
The TrustchainCommunity ID got changed. But there should be several other peers by now?
With the v7.2 hitting our frontpage already some time ago, we have 100+ installs that are not finding eachother. A lot of people try out our latest release in hours. All communities are healthy after a while, except 1 :
![screenshot from 2019-01-31 20-23-23](https://user-images.githubusercontent.com/325224/52079539-16d6aa80-2596-11e9-8024-15670980ca96.png)
2 Computer will be online all night. If you can't find them, we have an issue.

**Comment by devos50** [comment]
Same issue here:

<img width="808" alt="b9a0253795ca4cd62a84f7dd5e144393" src="https://user-images.githubusercontent.com/1707075/52080271-a892e780-2597-11e9-8cb3-1e8f72416152.png">

The amount of peers in the `TrustChainCommunity` stays on zero, even after some time. This is not captured by our automated tests since they operate in the testnet and for some reason, can find other peers there (see [this screenshot](https://jenkins-ci.tribler.org/job/Deployment-Tests/job/Run_Tribler_Release/job/Run_Mac_Mini/176/artifact/output/ipv8_overlays.png)).

While this is a bug, having no peers in the TrustChain community is not affecting most of the core functionality of Tribler. Payouts when downloading anonymously are fully handled by the `TriblerTunnelCommunity`. Direct payouts first do a DHT lookup and contact the peer with the looked up IP address/port directly. The only thing is that users are not actively collecting TrustChain records from other peers now.

My best guess would be that this has something to with the IPv8 trackers not properly introducing peers to each other. This is an issue we should have captured before releasing. In the future, I would suggest to also run the application tester for a short period of time on the 'live' network (so not only on the testnet), plot the IPv8 overlay statistics and check whether Tribler is able to correctly find other peers.

**Comment by qstokkink** [comment]
Cannot reproduce, I have peers in several seconds running on the `release-7.2.0` branch:

![afbeelding](https://user-images.githubusercontent.com/3630389/52082159-50aaaf80-259c-11e9-9167-d54f0fedd3a2.png)

Does the issue persist? Do you have any log files?

**Comment by synctext** [comment]
It does somewhat find peers after 105 minute, but not many:
![screenshot from 2019-01-31 21-58-36](https://user-images.githubusercontent.com/325224/52084927-7dae9080-25a3-11e9-9f3c-5bae304ad425.png)

**Comment by synctext** [comment]
Please try to find the root cause of failing for a 7.2.1 if possible Fri.

**Comment by qstokkink** [comment]
Instantly up to 20 now:

![afbeelding](https://user-images.githubusercontent.com/3630389/52112396-a9af2d80-2606-11e9-9fba-a41ecd869685.png)

Maybe you guys are just too impatient? 😃

**Comment by xoriole** [comment]
@devos50 I've restarted all the tunnel_helpers in the leaseweb machines to latest devel commit

**Comment by devos50** [comment]
@qstokkink it seems you are in a different community, are you sure that you are running 7.2 released version (since my community ID is different). After a few minutes, I only have one other discovered peer.

**Comment by xoriole** [comment]
I also have a different community ID (`cdecca487745..`) and 7 peers in 13 mins.
![trustchain_ipv8_peers](https://user-images.githubusercontent.com/1442867/52115803-d287f080-260f-11e9-9f11-0ab107e711db.png)
![trustchain_uptime](https://user-images.githubusercontent.com/1442867/52115877-fd724480-260f-11e9-8afd-07d6b8e2be9c.png)

**Comment by Thomasedv** [comment]
If it helps any... I'm on the 7.2 release (Windows x64 version) and got 22 (up from 20 as it was a bit earlier) peers on the TrustChainCommunity. My master peer is the same as devos and synctext.

**Comment by qstokkink** [comment]
Hmm.. there are no punctures being sent, this is worrying:

![afbeelding](https://user-images.githubusercontent.com/3630389/52118123-cb63e100-2615-11e9-932e-c821855970cf.png)

You can compare this to a healthy overlay:

![afbeelding](https://user-images.githubusercontent.com/3630389/52118174-f0585400-2615-11e9-89b6-c2a85f8c04e9.png)

**Comment by qstokkink** [comment]
However, of course, `TrustChainTestnetCommunity` is perfectly fine.

**Comment by qstokkink** [comment]
Found it:

https://github.com/Tribler/py-ipv8/pull/422/commits/135e93979aee921768fcf3b42850e27fdf0c267a

This would affect any TrustChainCommunity subclasses as well.

**Comment by qstokkink** [comment]
The fix was included in #4193 and should be part of the next release.

**Comment by synctext** [comment]
Apologies for bringing the bad news (7.2.1 after 13h uptime):
![screenshot from 2019-02-02 11-07-48](https://user-images.githubusercontent.com/325224/52162928-e8aeb300-26da-11e9-8100-4182577418d3.png)

**Comment by qstokkink** [comment]
I might actually be wrong, cannot reproduce.

## Isolated DHTDiscoveryCommunity (Dispersy style)
![afbeelding](https://user-images.githubusercontent.com/3630389/52163375-6bd30780-26e1-11e9-9bba-026cd9903348.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52163378-81e0c800-26e1-11e9-8e39-3b265a76e996.png)

## Non-Isolated DHTDiscoveryCommunity (IPv8-shared-pool style)
![afbeelding](https://user-images.githubusercontent.com/3630389/52163538-ab9aee80-26e3-11e9-87cc-d46a20f4da4a.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52163540-b48bc000-26e3-11e9-85bd-f8be0a88b57e.png)

**Comment by qstokkink** [comment]
Still cannot reproduce. Second idea: maybe this is due to this version still sharing a peer pool with Dispersy?

**Comment by qstokkink** [comment]
Still cannot reproduce after almost an hour:

![afbeelding](https://user-images.githubusercontent.com/3630389/52163903-6d083280-26e9-11e9-98fe-d2d78e395a80.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52163907-76919a80-26e9-11e9-8686-26b3f6f16b3b.png)

**Comment by synctext** [comment]
It seems to very slowly increase with uptime, see 15 hours sample on Linux. Note this is a fully connectable peer!
![screenshot from 2019-02-02 13-02-08](https://user-images.githubusercontent.com/325224/52163985-ea807280-26ea-11e9-9180-7eb3fb6bef34.png)

**Comment by qstokkink** [comment]
Over two hours in:

![afbeelding](https://user-images.githubusercontent.com/3630389/52164562-a180ec00-26f3-11e9-8b4f-9f2af6c61597.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52164564-a8a7fa00-26f3-11e9-8c2e-2b412da899c2.png)

Are we sure the correct IPv8 version was shipped with `7.2.1`?

**Comment by qstokkink** [comment]
Just over 3 hours in, still nothing out of the ordinary.

![afbeelding](https://user-images.githubusercontent.com/3630389/52165000-160b5900-26fb-11e9-8618-ff3107f2032d.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52165005-1f94c100-26fb-11e9-8342-a03185723d52.png)

**Comment by qstokkink** [comment]
Status update 4 hours in, still nothing:

![afbeelding](https://user-images.githubusercontent.com/3630389/52165590-21fb1900-2703-11e9-869b-c71c75f9e418.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52165597-2aebea80-2703-11e9-88b1-a739860eafda.png)

**Comment by qstokkink** [comment]
Captain's log, 5 hours in:

![afbeelding](https://user-images.githubusercontent.com/3630389/52166283-83bf8100-270b-11e9-8682-7566eee7b36a.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52166286-8cb05280-270b-11e9-936d-a5b73b9df265.png)

**Comment by qstokkink** [comment]
Calling it a day, can't reproduce:

![afbeelding](https://user-images.githubusercontent.com/3630389/52167038-2c261300-2715-11e9-82f3-af160d0f224e.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52167033-20d2e780-2715-11e9-931e-8aec7eeb8994.png)

**Comment by xoriole** [comment]
I'm sure the latest IPv8 version is shipped with 7.2.1. I just started running Tribler today, will see how it'll progress overnight.

**Comment by synctext** [comment]
Frontpage .deb of latest....

**Comment by qstokkink** [comment]
So far I've ran from source, I'll try the frontpage `.deb` tomorrow.

**Comment by xoriole** [comment]
My peer count in 12h 49m
![tribler-debug1](https://user-images.githubusercontent.com/1442867/52173230-91fbb480-2780-11e9-8932-5c4c49560ab8.png)
![tribler-debug2](https://user-images.githubusercontent.com/1442867/52173234-a8a20b80-2780-11e9-8fbf-b99492f2d3c1.png)

**Comment by qstokkink** [comment]
Hmm.. there are two commonalities between your screenshots.

 1. The Dispersy/DiscoveryCommunity peer count is well over 200. Given the amount of IPv8 and Dispersy communities, the peer count should not be that high.
 2. The DHTDiscoveryCommunity and TriblerTunnelCommunity are pulling ahead. This suggests that there may be some peer inflation in those communities.

My best guess right now is that this (recently added) line is causing problems in IPv8/Dispersy dual stack mode:

https://github.com/Tribler/tribler/blob/8ed2b916a118d689bd87766fb4041d7343cdcae4/Tribler/Core/APIImplementation/LaunchManyCore.py#L246

Before we added this line, the Dispersy community would be completely separate from IPv8. Now Dispersy peers may be bleeding into IPv8.

The solution would be to either:
 a. Leave out this line while in dual stack mode and make IPv8 less efficient (putting it [in this if-statement](https://github.com/Tribler/tribler/blob/8ed2b916a118d689bd87766fb4041d7343cdcae4/Tribler/Core/APIImplementation/LaunchManyCore.py#L248)).
 b. Make a special strategy to patch up peers coming in from Dispersy and leave them out of IPv8.

I think it would be best to use _(a)_ for now. When Dispersy is finally gone, we can get more efficiency.

**Comment by qstokkink** [comment]
If my intuition is correct, we could make it visible by periodically `clear()`-ing the Dispery `DiscoveryCommunity._candidates`. That should explode the peer count in IPv8.

**Comment by synctext** [comment]
Bleeding of Dispersy..
Possibly relevant is that I've also subscribed to a single popular Dispersy channel.

**Comment by qstokkink** [comment]
So.. I supercharged Dispersy's walking and I got over 450 peers which aren't going away in IPv8:

![afbeelding](https://user-images.githubusercontent.com/3630389/52174891-fb88bc80-279a-11e9-9b68-4302377c05a3.png)

It was Dispersy all along. I'll have to come up with a patch to stop this Dispersy fueled infinite growth.

**Comment by qstokkink** [comment]
I patched up the peer explosion caused by Dispersy. Still seems like the DHT and Tunnel communities are doing something fishy though.

**Comment by qstokkink** [comment]
With the Dispersy fixup, the DHT and Tunnel communities are also behaving normally again it seems.

**Comment by synctext** [comment]
Could you devise a minimal invasive patch? One that we can release without any risk for 7.2.2? (Post-weekend..)

**Comment by qstokkink** [comment]
😨 there is _also_ an issue with the DHT community pulling in too many peers. I'll handle that in IPv8 though, this issue is already too long.

**Comment by Dmole** [comment]
Good thing github has release notifications now so we don't need to keep checking back for when Tribler will become usable again... In case something is broken in 7.2.2, what was the last good version? (In case anyone wants to mine points without wasting electricity for fun)

**Comment by qstokkink** [comment]
The version not suffering from this particular version is 7.0.2. Do note that points do not carry over between these versions.

**Comment by synctext** [comment]
Gentle reminder that 7.2.1 on our frontpage after nearly 6 days uptime shows (recommending a 7.2.2):
![screenshot from 2019-02-07 14-25-47](https://user-images.githubusercontent.com/325224/52414344-547e8a80-2ae4-11e9-91e5-6f53d0310656.png)
It's the miracle of February that the CPU is not overloaded. This shatters my flawed belief that overlay neighbors lists should not be too big. For security, integrity and defense against eclipse attack we should move IPv8 to a ridiculously high 250 peers (insert master thesis on "NAT timeout detection"). For instance, [hijacking 37 BGP prefixes gives you control of Bitcoin](https://vanbever.eu/pdfs/vanbever_bitcoin_arxiv_2016.pdf).
![screenshot from 2019-02-07 14-26-00](https://user-images.githubusercontent.com/325224/52414669-3c5b3b00-2ae5-11e9-8e41-89ed8c444266.png)

**Comment by qstokkink** [comment]
We now have a job to show time-until-introduction for each of the isolated default communities in IPv8:
https://jenkins-ci.tribler.org/job/ipv8/job/performance_peer_discovery_defaults/plot/

![afbeelding](https://user-images.githubusercontent.com/3630389/52422224-ef339500-2af5-11e9-8e3b-75f41dc5b704.png)

**Comment by synctext** [comment]
Act III of the _the infinite growing overlay_.  Running the latest 7.2.2 for one night. The mystery deepens even further:
![img_20190209_130844](https://user-images.githubusercontent.com/325224/52520714-0b0d7700-2c6d-11e9-86d3-1178d62ae1da.jpg)

**Comment by qstokkink** [comment]
What in the world.. could you try disabling the DHTDiscoveryCommunity and/or the TriblerTunnelCommunity and see if it still happens?

**Comment by qstokkink** [comment]
At this point I my best theory is that there is something wrong with your environment.

**Comment by synctext** [comment]
Standard Mac... Reproducing on Ubuntu overnight..

**Comment by Dmole** [comment]
```
grep DEFAULT_MAX_PEERS /usr/share/tribler/Tribler/pyipv8/ipv8/community.py
DEFAULT_MAX_PEERS = 30
```
yet the DSDC climes to 133 within 13 minutes, other overlays are ~20.

**Comment by qstokkink** [comment]
@Dmole great, perfectly healthy. All overlays except the DSDC should fluctuate between 20~30, sporadically going over and under. The DSDC serves to monitor all other communities, but should never go over 180.

**Comment by Dmole** [comment]
To help identify peer count issues in the future (and in non-standard environments) maybe add that info to the debug UI by listing "current/max" in each row (150/180, 20/30, etc).

**Comment by qstokkink** [comment]
Perhaps we could color the output? Yellow for <20, green for 20 ~ 30, red for > 30.

**Comment by qstokkink** [comment]
Something like this:
![afbeelding](https://user-images.githubusercontent.com/3630389/52532637-a1eb3980-2d28-11e9-8850-a27bed4e0e9a.png)

**Comment by devos50** [comment]
A quick inspection of the TrustChain crawler log suggests that our crawler is not walking through the network anymore, i.e. it cannot/refuses to reach parts of the network. This explains the collapsing statistics on explorer.tribler.org. Logs are filled with the following line:

```
Feb 10 11:52:57 trustchain-crawler twistd[28213]: 2019-02-10T11:52:57+0000 [stderr#error] 2019-02-10 11:52:57,179:INFO:Dropping introduction request from (220.124.165.175, 56725): too many peers!
```

The crawler used to refresh its peers and fetch TrustChain records of the exit nodes. For some reason, it is not doing this anymore :(

**Comment by synctext** [comment]
Tried these terminal commands on the Mac.  After 40h uptime the communities are getting close to 500 peers each. None of the terminal commands work. No .py files in that location or No module named ipv8/Tribler...
Good new: Ubuntu 7.2.2 is OK and stable overlay count.

**Comment by devos50** [comment]
The commands that @qstokkink linked are specific for Linux environments and do not work on Mac and Windows. It is not possible to inspect the bundled Python code since it is compiled to byte code and compressed by Pyinstaller.

**Comment by qstokkink** [comment]
@devos50 Oh sorry, forgot to edit the crawlers for the new setup. You need to set the `max_peers` to `-1` for the crawling Community instances.

**Comment by devos50** [comment]
Thanks, I will fix that in a moment 👍

**Comment by qstokkink** [comment]
![afbeelding](https://user-images.githubusercontent.com/3630389/52537161-a2ed8c80-2d63-11e9-8ed3-a820d3a0892a.png)

Technically this could still happen naturally; but it seems that the TriblerTunnelCommunity/DHTDiscoveryCommunity are pulling in too many peers. We should make a build without the DiscoveryCommunity's SimilarityRequests for testing.

**Comment by qstokkink** [comment]
Good news and bad news:

 - Good news: @synctext you are not personally cursed.
 - Bad news: I reproduced this.

![afbeelding](https://user-images.githubusercontent.com/3630389/52542893-d69cd680-2da4-11e9-8ff5-74248d105c06.png)

 - Good news: I can see that the following communities are misbehaving:
   - `DHTDiscoveryCommunity`/`TriblerTunnelCommunity` (which I grouped because they seem to work together)
   - `DualStackDiscoveryCommunity` (which could potentially be a side effect from the previous two)

**Comment by qstokkink** [comment]
![afbeelding](https://user-images.githubusercontent.com/3630389/52557633-4b493280-2df0-11e9-8ea1-fac95a725f55.png)

Hmm.. on my latest PR for IPv8 (https://github.com/Tribler/py-ipv8/pull/431) the DHT is still pulling ahead quite rapidly.

**Comment by qstokkink** [comment]
After a debug day with @egbertbouman, we assume this to be fixed (in the new release - when the IPv8 pointer is updated).

![afbeelding](https://user-images.githubusercontent.com/3630389/52574509-c3c5e880-2e1c-11e9-849d-14901d40476d.png)
![afbeelding](https://user-images.githubusercontent.com/3630389/52574540-d2140480-2e1c-11e9-8dcd-ccde1b597650.png)

**Comment by egbertbouman** [comment]
Most likely this was caused by punctures received by the IPv8 DiscoveryCommunity that are send by Dispersy. The `on_puncture` added these peers into the `verified_peers` list, thus causing the leak.
